clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1614,73 +1616,63 @@ mlir::Value ScalarExprEmitter::VisitUnaryLNot(const UnaryOperator *E) {
   llvm_unreachable("destination type for negation unary operator is NYI");
 }
 
+// Conversion from bool, integral, or floating-point to integral or
+// floating-point. Conversions involving other types are handled elsewhere.
+// Conversion to bool is handled elsewhere because that's a comparison against
+// zero, not a simple cast.
 mlir::Value ScalarExprEmitter::buildScalarCast(
     mlir::Value Src, QualType SrcType, QualType DstType, mlir::Type SrcTy,
     mlir::Type DstTy, ScalarConversionOpts Opts) {
-  // The Element types determine the type of cast to perform.
-  mlir::Type SrcElementTy;
-  mlir::Type DstElementTy;
-  QualType SrcElementType;
-  QualType DstElementType;
-  if (SrcType->isMatrixType() && DstType->isMatrixType()) {
-    llvm_unreachable("NYI");
-  } else {
-    assert(!SrcType->isMatrixType() && !DstType->isMatrixType() &&
-           "cannot cast between matrix and non-matrix types");
-    SrcElementTy = SrcTy;
-    DstElementTy = DstTy;
-    SrcElementType = SrcType;
-    DstElementType = DstType;
-  }
 
-  if (SrcElementTy.isa<mlir::IntegerType>() ||
-      DstElementTy.isa<mlir::IntegerType>())
+  if (SrcTy.isa<mlir::IntegerType>() || DstTy.isa<mlir::IntegerType>())
     llvm_unreachable("Obsolete code. Don't use mlir::IntegerType with CIR.");
 
-  if (SrcElementType->isBooleanType()) {
+  mlir::cir::CastKind CastKind = mlir::cir::CastKind(0);
```

Comment:
Since the enumeration isn't guaranteed to be stable, can you change this to an optional instead?

Reply:
Thanks for the suggestion.  That's an improvement.  (std::optional hasn't made into my usual bag of tricks yet.)